core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ImplementTableFunctionSource.java

```diff
@@ -265,18 +317,243 @@ private static NodeWithSymbols appendHelperSymbolsForSource(
+        sourceList = sourceList.stream()
+                .sorted(Comparator.comparingInt(source -> source.getProperties().isPruneWhenEmpty() ? -1 : 1))
+                .collect(toImmutableList());
+
+        NodeWithSymbols first = appendHelperSymbolsForSource(sourceList.get(0).getSource(), sourceList.get(0).getProperties(), partitionSizeSymbols, rowNumberSymbols, metadata, context);
+        NodeWithSymbols second = appendHelperSymbolsForSource(sourceList.get(1).getSource(), sourceList.get(1).getProperties(), partitionSizeSymbols, rowNumberSymbols, metadata, context);
+        JoinedNodes coPartitioned = coPartition(first, second, context);
+
+        for (int i = 2; i < sourceList.size(); i++) {
+            NodeWithSymbols coPartitionedWithSymbols = appendHelperSymbolsForCoPartitionedNodes(coPartitioned, partitionSizeSymbols, combinedRowNumberSymbols, combinedPartitionBySymbols, context);
+            NodeWithSymbols next = appendHelperSymbolsForSource(sourceList.get(i).getSource(), sourceList.get(i).getProperties(), partitionSizeSymbols, rowNumberSymbols, metadata, context);
+            coPartitioned = coPartition(coPartitionedWithSymbols, next, context);
+        }
+
+        return appendHelperSymbolsForCoPartitionedNodes(coPartitioned, partitionSizeSymbols, combinedRowNumberSymbols, combinedPartitionBySymbols, context);
+    }
+
+    private static JoinedNodes coPartition(NodeWithSymbols left, NodeWithSymbols right, Context context)
+    {
+        checkArgument(left.getPartitionBy().size() == right.getPartitionBy().size(), "co-partitioning lists do not match");
+
+        Expression leftRowNumber = left.getRowNumber().toSymbolReference();
+        Expression leftPartitionSize = left.getPartitionSize().toSymbolReference();
+        List<Expression> leftPartitionBy = left.getPartitionBy().stream()
+                .map(Symbol::toSymbolReference)
+                .collect(toImmutableList());
+        Expression rightRowNumber = right.getRowNumber().toSymbolReference();
+        Expression rightPartitionSize = right.getPartitionSize().toSymbolReference();
+        List<Expression> rightPartitionBy = right.getPartitionBy().stream()
+                .map(Symbol::toSymbolReference)
+                .collect(toImmutableList());
+
+        List<Expression> coPartitionConjuncts = Streams.zip(
+                        leftPartitionBy.stream(),
+                        rightPartitionBy.stream(),
+                        (leftColumn, rightColumn) -> new NotExpression(new ComparisonExpression(IS_DISTINCT_FROM, leftColumn, rightColumn)))
+                .collect(toImmutableList());
+        // In StatementAnalyzer we require that co-partitioned tables have non-empty partitioning column lists.
+        // Co-partitioning tables with empty partition by would be ineffective.
+        checkState(!coPartitionConjuncts.isEmpty(), "co-partitioned tables must have partitioning columns");
```

Comment:
Move this check to the top of this method (and use `!left.getPartitionBy().isEmpty()` instead).